libbeat/processors/dns/cache.go

```diff
@@ -0,0 +1,194 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package dns
+
+import (
+	"time"
+
+	"github.com/elastic/beats/libbeat/monitoring"
+)
+
+type ptrRecord struct {
+	host    string
+	expires time.Time
+}
+
+func (r ptrRecord) IsExpired(now time.Time) bool {
+	return now.After(r.expires)
+}
+
+type ptrCache struct {
+	data    map[string]ptrRecord
+	maxSize int
+}
+
+func (c *ptrCache) set(now time.Time, key string, ptr *PTR) {
+	if len(c.data) >= c.maxSize {
+		c.evict()
+	}
+
+	c.data[key] = ptrRecord{
+		host:    ptr.Host,
+		expires: now.Add(time.Duration(ptr.TTL) * time.Second),
+	}
+}
+
+func (c *ptrCache) evict() {
+	var key string
+	for k := range c.data {
+		key = k
+		break
+	}
+	delete(c.data, key)
+}
+
+func (c *ptrCache) get(now time.Time, key string) *PTR {
+	r, found := c.data[key]
+	if found && !r.IsExpired(now) {
+		return &PTR{r.host, uint32(r.expires.Sub(now) / time.Second)}
+	}
+	return nil
+}
+
+type failureRecord struct {
+	error
+	expires time.Time
+}
+
+func (r failureRecord) IsExpired(now time.Time) bool {
+	return now.After(r.expires)
+}
+
+type failureCache struct {
+	data       map[string]failureRecord
+	maxSize    int
+	failureTTL time.Duration
+	stats      cacheStats
+}
+
+func (c *failureCache) set(now time.Time, key string, err error) {
+	if len(c.data) >= c.maxSize {
+		c.evict()
+	}
+
+	c.data[key] = failureRecord{
+		error:   err,
+		expires: now.Add(c.failureTTL),
+	}
+}
+
+func (c *failureCache) evict() {
+	var key string
+	for k := range c.data {
+		key = k
+		break
+	}
+	delete(c.data, key)
+}
+
+func (c *failureCache) get(now time.Time, key string) error {
+	r, found := c.data[key]
+	if found && !r.IsExpired(now) {
+		return r.error
+	}
+	return nil
+}
+
+// PTRLookupCache is a cache for storing and retrieving the results of
+// reverse DNS queries. It caches the results of queries regardless of their
+// outcome (success or failure).
+type PTRLookupCache struct {
+	success    *ptrCache
+	failure    *failureCache
+	failureTTL time.Duration
+	resolver   PTRResolver
+	log        Logger
+	stats      cacheStats
+}
+
+type cacheStats struct {
+	Hit  *monitoring.Int
+	Miss *monitoring.Int
+}
+
+// Logger logs debug messages.
+type Logger interface {
+	Debugw(msg string, keysAndValues ...interface{})
+}
+
+// NewPTRLookupCache returns a new cache.
+func NewPTRLookupCache(reg *monitoring.Registry, l Logger, conf CacheConfig, resolver PTRResolver) *PTRLookupCache {
+	c := &PTRLookupCache{
+		success: &ptrCache{
+			data:    make(map[string]ptrRecord, conf.SuccessCache.InitialCapacity),
+			maxSize: max(100, max(conf.SuccessCache.InitialCapacity, conf.SuccessCache.MaxCapacity)),
```

Comment:
This behavior is a little weird. It essentially says you can't set a cache size `< 100`. I'm not so certain we want to get into the business of overriding user preferences silently. Shouldn't we fail to start and log a warning?